File: IELTS_Speaking/extract_from_docx/p4.go
package main

import (
	"context"
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
	"sync"

	"github.com/openai/openai-go"
	"github.com/openai/openai-go/option"
)

type dbRow struct {
	topic string
	seq   int
}

type spkTopic struct {
	id   string
	name string
}

const groupSize = 200

const promptTemplate = `我正在备考雅思口语。以下是我当前背诵的词伙词组列表，以及这些词组所属的分类：

%s

以下是可供匹配的雅思口语话题列表（topic_id, topic_name）：
%s

请根据词组的语义内容，找出与这些词组最相关的口语话题。只返回匹配的 topic_id，用空格分隔，放在 <spk_topic_ids> 标签里。
如果没有匹配的话题，就返回 <spk_topic_ids></spk_topic_ids>。

%s`

func main() {
	// --- 1. 读取 db.csv ---
	dbRows := readDB("db.csv")

	// --- 2. 读取 speaking.csv，去重得到 spk_topics (topic_id -> topic_name) ---
	spkTopics := readSpeaking("speaking.csv")
	spkTopicsStr := formatSpkTopics(spkTopics)

	// --- 3. 按 seq 分组，得到每组的 distinct topic ---
	groups := groupBySeq(dbRows)
	sortedKeys := make([]int, 0, len(groups))
	for k := range groups {
		sortedKeys = append(sortedKeys, k)
	}
	sort.Ints(sortedKeys)

	// --- 4. 初始化 Client（和 reference.go 一样）---
	client := openai.NewClient(
		option.WithAPIKey(os.Getenv("DASHSCOPE_API_KEY")),
		option.WithBaseURL("https://dashscope.aliyuncs.com/compatible-mode/v1"),
	)

	// --- 5. 并发调用 LLM ---
	type groupResult struct {
		groupNumber int
		topicIDs    []string
	}
	results := make(map[int][]string)
	var mu sync.Mutex
	var wg sync.WaitGroup

	type job struct {
		groupNumber int
		topics      []string
	}
	jobs := make(chan job, len(sortedKeys))

	// 下发所有任务
	for _, gn := range sortedKeys {
		jobs <- job{groupNumber: gn, topics: groups[gn]}
	}
	close(jobs)

	// 启动 worker 池（10 个并发）
	numWorkers := 10
	for w := 1; w <= numWorkers; w++ {
		wg.Add(1)
		go func(wid int, client *openai.Client, spkStr string) {
			defer wg.Done()
			for jb := range jobs {
				distinctStr := formatDistinctTopics(jb.topics)
				require := "只返回 topic_id，用空格分隔，放在 <spk_topic_ids> 标签里。不要输出任何其他内容。"
				prompt := fmt.Sprintf(promptTemplate, distinctStr, spkStr, require)

				fmt.Printf("[Worker %d] Group %d (%d topics)... ", wid, jb.groupNumber, len(jb.topics))

				chatCompletion, err := client.Chat.Completions.New(
					context.TODO(), openai.ChatCompletionNewParams{
						Messages: []openai.ChatCompletionMessageParamUnion{
							openai.UserMessage(prompt),
						},
						Model: "qwen3.6-plus",
					},
				)
				if err != nil {
					log.Printf("[Worker %d] Group %d 调用失败: %v\n", wid, jb.groupNumber, err)
					mu.Lock()
					results[jb.groupNumber] = nil
					mu.Unlock()
					continue
				}

				content := chatCompletion.Choices[0].Message.Content
				ids := extractTopicIDs(content)
				fmt.Printf("匹配 %d 个 topic\n", len(ids))

				mu.Lock()
				results[jb.groupNumber] = ids
				mu.Unlock()
			}
		}(w, &client, spkTopicsStr)
	}

	wg.Wait()
	fmt.Println("\n所有组处理完毕。")

	// --- 6. 按 group_number 排序后输出 ---
	topicIDToName := buildTopicIDMap(spkTopics)

	out, err := os.Create("group_topic.csv")
	if err != nil {
		log.Fatalf("创建输出文件失败: %v", err)
	}
	defer out.Close()

	writer := csv.NewWriter(out)
	defer writer.Flush()

	writer.Write([]string{"group_number", "topic_id", "topic_name"})

	total := 0
	for _, gn := range sortedKeys {
		tids := results[gn]
		for _, tid := range tids {
			tname := topicIDToName[tid]
			writer.Write([]string{
				strconv.Itoa(gn),
				tid,
				tname,
			})
			total++
		}
	}

	fmt.Printf("完成！共 %d 组，%d 个匹配关系，已保存至 group_topic.csv\n", len(sortedKeys), total)
}

// extractTopicIDs 从 LLM 返回中提取 <spk_topic_ids> 标签内的内容
func extractTopicIDs(content string) []string {
	start := strings.Index(content, "<spk_topic_ids>")
	end := strings.Index(content, "</spk_topic_ids>")
	if start == -1 || end == -1 {
		// fallback: 直接从内容中提取数字
		var ids []string
		for _, token := range strings.Fields(content) {
			if _, err := strconv.Atoi(token); err == nil {
				ids = append(ids, token)
			}
		}
		return ids
	}
	body := content[start+len("<spk_topic_ids>") : end]
	var ids []string
	for _, token := range strings.Fields(body) {
		token = strings.TrimSpace(token)
		if token == "" {
			continue
		}
		if _, err := strconv.Atoi(token); err == nil {
			ids = append(ids, token)
		}
	}
	return ids
}

func formatDistinctTopics(topics []string) string {
	var sb strings.Builder
	for _, t := range topics {
		sb.WriteString(fmt.Sprintf("  - %s\n", t))
	}
	return sb.String()
}

func formatSpkTopics(topics []spkTopic) string {
	var sb strings.Builder
	for _, t := range topics {
		sb.WriteString(fmt.Sprintf("  topic_id=%s, topic_name=%s\n", t.id, t.name))
	}
	return sb.String()
}

func buildTopicIDMap(topics []spkTopic) map[string]string {
	m := make(map[string]string)
	for _, t := range topics {
		if _, ok := m[t.id]; !ok {
			m[t.id] = t.name
		}
	}
	return m
}

func groupBySeq(rows []dbRow) map[int][]string {
	groups := make(map[int]map[string]bool)
	for _, row := range rows {
		groupNum := (row.seq-1)/groupSize + 1
		if groups[groupNum] == nil {
			groups[groupNum] = make(map[string]bool)
		}
		if row.topic != "" {
			groups[groupNum][row.topic] = true
		}
	}
	result := make(map[int][]string)
	for k, v := range groups {
		var topics []string
		for t := range v {
			topics = append(topics, t)
		}
		result[k] = topics
	}
	return result
}

func readDB(path string) []dbRow {
	f, err := os.Open(path)
	if err != nil {
		log.Fatalf("无法打开 %s: %v", path, err)
	}
	defer f.Close()

	reader := csv.NewReader(f)
	records, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("读取 %s 失败: %v", path, err)
	}

	var rows []dbRow
	for i, record := range records {
		if i == 0 {
			continue // skip header
		}
		if len(record) < 4 {
			continue
		}
		seq, err := strconv.Atoi(strings.TrimSpace(record[3]))
		if err != nil {
			log.Printf("跳过第 %d 行, seq 解析失败: %v\n", i, err)
			continue
		}
		rows = append(rows, dbRow{
			topic: strings.TrimSpace(record[0]),
			seq:   seq,
		})
	}
	return rows
}

func readSpeaking(path string) []spkTopic {
	f, err := os.Open(path)
	if err != nil {
		log.Fatalf("无法打开 %s: %v", path, err)
	}
	defer f.Close()

	reader := csv.NewReader(f)
	reader.LazyQuotes = true
	records, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("读取 %s 失败: %v", path, err)
	}

	seen := make(map[string]bool)
	var topics []spkTopic
	for i, record := range records {
		if i == 0 {
			continue
		}
		if len(record) < 5 {
			continue
		}
		tid := strings.TrimSpace(record[1])
		tname := strings.TrimSpace(record[3])
		key := tname
		if seen[key] {
			continue
		}
		seen[key] = true
		topics = append(topics, spkTopic{id: tid, name: tname})
	}
	return topics
}

File: IELTS_Speaking/extract_from_docx/p4.py
import csv
import os
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from openai import OpenAI

GROUP_SIZE = 200

PROMPT_TEMPLATE = """\
我正在备考雅思口语。以下是我当前背诵的词伙词组列表，以及这些词组所属的雅思话题分类：

{db_distinct_topics}

以下是可供匹配的雅思口语话题列表：
{spk_topics}

请根据词组的内容，找出与这个词组最相关的口语话题。每个话题返回 topic_id 即可，用空格隔开多个 topic_id，放在 <spk_topic_ids> 标签里。
如果没有匹配的话题，就返回 <spk_topic_ids></spk_topic_ids>。

require: {require}
"""


def main():
    # --- 1. 读取 db.csv ---
    db_rows = []
    with open("db.csv", "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            db_rows.append(row)

    # --- 2. 读取 speaking.csv，构造 spk_topics ---
    spk_rows = []
    with open("speaking.csv", "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            spk_rows.append(row)

    # 去重: topic_id -> topic_name
    seen = {}
    for row in spk_rows:
        tid = row["topic_id"].strip()
        tname = row["topic_name"].strip()
        if tid not in seen:
            seen[tid] = tname
    spk_topics = [(tid, tname) for tid, tname in seen.items()]
    spk_topics_str = "\n".join(f"  topic_id={tid}, topic_name={tname}" for tid, tname in spk_topics)

    # --- 3. 按 seq 分组 ---
    groups = {}
    for row in db_rows:
        seq = int(row["seq"].strip())
        group_number = (seq - 1) // GROUP_SIZE + 1
        if group_number not in groups:
            groups[group_number] = set()
        topic = row["topic"].strip()
        if topic:
            groups[group_number].add(topic)

    # --- 4. 构造 LLM 调用 ---
    client = OpenAI(
        api_key=os.environ["DASHSCOPE_API_KEY"],
        base_url="https://dashscope.aliyuncs.com/compatible-mode/v1",
    )

    results = {}  # group_number -> list of topic_ids

    def call_llm(group_number, distinct_topics):
        require = f"请只返回 <spk_topic_ids> 标签包裹的 topic_id 数字，用空格分隔。不要输出任何其他内容。"
        topics_list = "\n".join(f"  - {t}" for t in sorted(distinct_topics))
        prompt = PROMPT_TEMPLATE.format(
            db_distinct_topics=topics_list,
            spk_topics=spk_topics_str,
            require=require,
        )

        print(f"[Group {group_number}] 调用 LLM... ({len(distinct_topics)} distinct topics)")

        response = client.chat.completions.create(
            model="qwen3.6-plus",
            messages=[{"role": "user", "content": prompt}],
        )

        content = response.choices[0].message.content.strip()
        print(f"[Group {group_number}] 返回: {content}")

        # 提取 <spk_topic_ids> 标签里的内容
        import re
        m = re.search(r"<spk_topic_ids>(.*?)</spk_topic_ids>", content, re.DOTALL)
        if m:
            ids_text = m.group(1).strip()
            if ids_text:
                ids = [tid.strip() for tid in ids_text.split() if tid.strip().isdigit()]
            else:
                ids = []
        else:
            # fallback: 尝试直接从内容中提取数字
            ids = [s for s in content.split() if s.strip().isdigit()]

        results[group_number] = ids

    # 顺序调用，避免并发超限
    sorted_groups = sorted(groups.items())
    for group_number, distinct_topics in sorted_groups:
        try:
            call_llm(group_number, distinct_topics)
        except Exception as e:
            print(f"[Group {group_number}] 错误: {e}")
            results[group_number] = []
        time.sleep(1)

    # --- 5. 输出 group_topic.csv ---
    with open("group_topic.csv", "w", encoding="utf-8-sig", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(["group_number", "topic_id", "topic_name"])

        topic_id_to_name = {tid: tname for tid, tname in spk_topics}

        for group_number in sorted(results.keys()):
            for tid in sorted(results[group_number], key=int):
                tname = topic_id_to_name.get(tid, "Unknown")
                writer.writerow([group_number, tid, tname])

    total = sum(len(ids) for ids in results.values())
    print(f"\n完成！共 {len(results)} 组，{total} 个匹配，已保存至 group_topic.csv")


if __name__ == "__main__":
    main()
